### PR TITLE
sdkwsgi update for list fcp tempate API

### PR DESCRIPTION
**Pull request by haolp**
```
>>> con.send_request('get_fcp_templates',template_id_list=['36439338-db14-11ec-bb41-0201018b1dd2','12345678'])
{'overallRC': 0, 'modID': None, 'rc': 0, 'rs': 0, 'errmsg': '', 'output': 
{
    'fcp_templates': [
        {'id': '12345678', 'name': 'templatet1', 
         'description': 'test1', 'is_default': 0, 
         'sp_name': ['sp3']
        }, 
        {'id': '36439338-db14-11ec-bb41-0201018b1dd2', 
         'name': 'default_template', 'description': 'This is Default template', 
         'is_default': 1, 
         'sp_name': ['v7k60', 'sp4']
        }
    ]
}}
```
```
>>> pp(con.send_request('get_fcp_templates',default_sp_list=['v7k60','sp3']))
{'errmsg': '',
 'modID': None,
 'output': {'fcp_templates': [{'description': 'This is Default template',
                               'id': '36439338-db14-11ec-bb41-0201018b1dd2',
                               'is_default': 1,
                               'name': 'default_template',
                               'sp_name': ['v7k60']},
                              {'description': 'test1',
                               'id': '12345678',
                               'is_default': 0,
                               'name': 'templatet1',
                               'sp_name': ['sp3']}]},
 'overallRC': 0,
 'rc': 0,
 'rs': 0}
```
```
>>> pp(con.send_request('get_fcp_templates',host_default=True))
{'errmsg': '',
 'modID': None,
 'output': {'fcp_templates': [{'description': 'This is Default template',
                               'id': '36439338-db14-11ec-bb41-0201018b1dd2',
                               'is_default': 1,
                               'name': 'default_template',
                               'sp_name': ['sp4', 'v7k60']}]},
 'overallRC': 0,
 'rc': 0,
 'rs': 0}

```
```
>>> con.send_request('get_fcp_templates',assigner_id='HLP0000B')
{'overallRC': 0, 'modID': None, 'rc': 0, 'rs': 0, 'errmsg': '', 'output': 
{'fcp_templates': [
    {'id': '36439338-db14-11ec-bb41-0201018b1dd2', 
     'name': 'default_template', 'description': 'This is Default template', 
     'is_default': 1, 
     'sp_name': ['sp4', 'v7k60']}]}}
```
```
>>> pp(con.send_request('get_fcp_templates',default_sp_list=['all']))
{'errmsg': '',
 'modID': None,
 'output': {'fcp_templates': [{'description': 'This is Default template',
                               'id': '36439338-db14-11ec-bb41-0201018b1dd2',
                               'is_default': 1,
                               'name': 'default_template',
                               'sp_name': ['v7k60', 'sp4']},
                              {'description': 'just for test',
                               'id': '36439338-db14-11ec-bb41-0201018b1dd3',
                               'is_default': 0,
                               'name': 'test_template',
                               'sp_name': ['ds8k60c1']},
                              {'description': 'test1',
                               'id': '12345678',
                               'is_default': 0,
                               'name': 'templatet1',
                               'sp_name': ['sp3']}]},
 'overallRC': 0,
 'rc': 0,
 'rs': 0}
```

```
>>> pp(con.send_request('get_fcp_templates_details',template_id_list=['36439338-db14-11ec-bb41-0201018b1dd2'], raw=True))
{'errmsg': '',
 'modID': None,
 'output': {'fcp_templates': [{'description': 'This is Default template',
                               'id': '36439338-db14-11ec-bb41-0201018b1dd2',
                               'is_default': 1,
                               'name': 'default_template',
                               'raw': {'0': [['1a0f',
                                              '36439338-db14-11ec-bb41-0201018b1dd2',
                                              'HLP0000B',
                                              0,
                                              0,
                                              'c05076de3300038b',
                                              'c05076de33002e41',
                                              '27',
                                              'free',
                                              'none',
                                              '36439338-db14-11ec-bb41-0201018b1dd2'],
                                             ['1a0e',
                                              '36439338-db14-11ec-bb41-0201018b1dd2',
                                              '',
                                              0,
                                              0,
                                              'c05076de330003a2',
                                              'c05076de33002e41',
                                              '27',
                                              'free',
                                              'none',
                                              '36439338-db14-11ec-bb41-0201018b1dd2']],
                                       '1': [['1c0d',
                                              '36439338-db14-11ec-bb41-0201018b1dd2',
                                              '',
                                              0,
                                              0,
                                              'c05076de33000353',
                                              'c05076de33002641',
                                              '32',
                                              'free',
                                              'none',
                                              '36439338-db14-11ec-bb41-0201018b1dd2']]},
                               'sp_name': ['v7k60', 'sp4'],
                               'statistics': {'0': {'CHPIDs': {'27': '1A0E - '
                                                                     '1A0F'},
                                                    'allocated': '',
                                                    'allocated_but_free': '',
                                                    'available': '1A0E - 1A0F',
                                                    'connection_only': '',
                                                    'notfound': '',
                                                    'offline': '',
                                                    'reserve_only': '',
                                                    'total': '1A0E - 1A0F',
                                                    'unallocated_but_active': []},
                                              '1': {'CHPIDs': {'32': '1C0D'},
                                                    'allocated': '',
                                                    'allocated_but_free': '',
                                                    'available': '1C0D',
                                                    'connection_only': '',
                                                    'notfound': '',
                                                    'offline': '',
                                                    'reserve_only': '',
                                                    'total': '1C0D',
                                                    'unallocated_but_active': []}}}]},
 'overallRC': 0,
 'rc': 0,
 'rs': 0}

```

### CODE REVIEW

File: zvmsdk/sdkwsgi/handler.py [hunk 1]
```diff
@@ -59,6 +59,12 @@
     }),
      ('/volumes/fcp', {
         'GET': volume.get_all_fcp_usage,
+    }),
+     ('/volumes/fcp_templates', {
```
Comment (IBMCaoBiao) [inline zvmsdk/sdkwsgi/handler.py:63]: can we use `fcptemplates`?

File: zvmsdk/sdkwsgi/handler.py [hunk 2]
```diff
@@ -59,6 +59,12 @@
...
+        'GET': volume.get_fcp_templates,
+    }),
+     ('/volumes/fcp_templates_details', {
```
Comment (IBMCaoBiao) [inline zvmsdk/sdkwsgi/handler.py:66]: no need to use separate url for this? I think we can use a parameter in url, like:
```
volumes/fcptemplates?detail=True
```

File: zvmconnector/restclient.py [hunk 3]
```diff
@@ -462,6 +462,42 @@ def req_get_all_fcp_usage(start_index, *args, **kwargs):
...
 
 
+def req_get_fcp_templates(start_index, *args, **kwargs):
+    url = '/volumes/fcptemplates'
+    template_id_list = kwargs.get('template_id_list', None)
+    host_default = kwargs.get('host_default', False)
+    default_sp_list = kwargs.get('default_sp_list', None)
+    assigner_id = kwargs.get('assigner_id', None)
+
+    if template_id_list:
```
Comment (jackydalong) [inline zvmconnector/restclient.py:472]: Do we need to keep the same order with this comment https://github.com/openmainframeproject/feilong/pull/608/files#r886233006 ?

File: zvmsdk/sdkwsgi/validation/parameter_types.py [hunk 4]
```diff
@@ -343,6 +343,23 @@ def valid_char(char):
...
 
+fcp_template_id_list = {
+    'items': {
+        'type': 'string',
+        'minLength': 1,
+    },
+    'type': 'array'
+}
+
+fcp_template_filters = {
```
Comment (jackydalong) [inline zvmsdk/sdkwsgi/validation/parameter_types.py:354]: Is this `fcp_template_filters` still needed ?

File: zvmsdk/sdkwsgi/handlers/volume.py [hunk 5]
```diff
@@ -247,3 +264,88 @@ def _get_all_fcp_usage(req, userid, raw, statistics, sync_with_zvm):
...
+        action = get_action()
+        return action.get_fcp_templates(req, template_id_list, assigner_id,
+                                        default_sp_list, host_default)
+
+    if 'template_id_list' in req.GET.keys():
+        template_id_list = req.GET['template_id_list']
+    else:
+        template_id_list = None
+
+    if 'assigner_id' in req.GET.keys():
```
Comment (jackydalong) [inline zvmsdk/sdkwsgi/handlers/volume.py:283]: ditto

File: zvmsdk/sdkwsgi/handlers/volume.py [hunk 6]
```diff
@@ -247,3 +264,88 @@ def _get_all_fcp_usage(req, userid, raw, statistics, sync_with_zvm):
...
+        assigner_id = req.GET['assigner_id']
+    else:
+        assigner_id = None
+
+    if 'host_default' in req.GET.keys():
```
Comment (jackydalong) [inline zvmsdk/sdkwsgi/handlers/volume.py:288]: ditto

File: zvmsdk/sdkwsgi/handlers/volume.py [hunk 7]
```diff
@@ -247,3 +264,88 @@ def _get_all_fcp_usage(req, userid, raw, statistics, sync_with_zvm):
...
+        host_default = req.GET['host_default']
+    else:
+        host_default = False
+
+    if 'default_sp_list' in req.GET.keys():
```
Comment (jackydalong) [inline zvmsdk/sdkwsgi/handlers/volume.py:293]: ditto

File: zvmconnector/restclient.py (hunk 3 above)
Comment (haolp) [reply zvmconnector/restclient.py:472]: updated, thanks.

File: zvmsdk/sdkwsgi/validation/parameter_types.py (hunk 4 above)
Comment (haolp) [reply zvmsdk/sdkwsgi/validation/parameter_types.py:354]: no need, removed, thanks.

File: zvmconnector/restclient.py [hunk 8]
```diff
@@ -462,6 +462,42 @@ def req_get_all_fcp_usage(start_index, *args, **kwargs):
...
+        url += "?template_id_list=%s" % template_id_list
+    elif assigner_id:
+        url += "?assigner_id=%s" % assigner_id
+    elif host_default:
+        url += "?host_default=%s" % host_default
+    elif default_sp_list:
+        url += "?default_sp_list=%s" % default_sp_list
```
[the remark this answers is not in the document]
Comment (haolp) [reply zvmconnector/restclient.py:479]: no, they can not exist at the same time.

File: zvmsdk/sdkwsgi/handlers/volume.py [hunk 9]
```diff
@@ -276,4 +293,65 @@ def _create_fcp_template(req):
...
+    def _get_fcp_templates(req, template_id_list, assigner_id,
+                           default_sp_list, host_default):
+        action = get_action()
+        return action.get_fcp_templates(req, template_id_list, assigner_id,
+                                        default_sp_list, host_default)
+
+    template_id_list = req.GET.get('template_id_list', None)
+    assigner_id = req.GET.get('assigner_id', None)
+    default_sp_list = req.GET.get('default_sp_list', None)
+    host_default = req.GET.get('host_default', None)
```
Comment (jackydalong) [inline zvmsdk/sdkwsgi/handlers/volume.py:307]: host_default should be False rather than None

**Review by jackydalong — APPROVED** [review]
LGTM

File: zvmsdk/sdkwsgi/handlers/volume.py (hunk 9 above)
Comment (haolp) [reply zvmsdk/sdkwsgi/handlers/volume.py]: updated, thanks.

File: zvmsdk/sdkwsgi/handler.py [hunk 10]
```diff
@@ -62,6 +62,12 @@
     }),
     ('/volumes/fcptemplates', {
         'POST': volume.create_fcp_template,
+    }),
+     ('/volumes/fcptemplates', {
```
Comment (IBMCaoBiao) [inline zvmsdk/sdkwsgi/handler.py:66]: we can combine GET and above POST together.

Comment (haolp) [reply zvmsdk/sdkwsgi/handler.py]: updated, thanks.

**Pull request merged**